ts/src/huobi.ts

```diff
@@ -4246,134 +4250,119 @@ export default class huobi extends Exchange {
+                if (price !== undefined) {
+                    request['sl_order_price'] = this.priceToPrecision (symbol, price);
+                }
+            } else {
+                request['tp_order_price_type'] = type;
+                request['tp_trigger_price'] = this.priceToPrecision (symbol, takeProfitTriggerPrice);
+                if (price !== undefined) {
+                    request['tp_order_price'] = this.priceToPrecision (symbol, price);
+                }
+            }
+        } else {
+            const clientOrderId = this.safeString2 (params, 'client_order_id', 'clientOrderId');
+            if (clientOrderId !== undefined) {
+                request['client_order_id'] = clientOrderId;
+                params = this.omit (params, [ 'client_order_id', 'clientOrderId' ]);
+            }
+            if (type === 'limit' || type === 'ioc' || type === 'fok' || type === 'post_only') {
+                request['price'] = this.priceToPrecision (symbol, price);
+            }
+        }
+        if (!isStopLossTriggerOrder && !isTakeProfitTriggerOrder) {
+            const leverRate = this.safeInteger2 (params, 'leverRate', 'lever_rate', 1);
+            const reduceOnly = this.safeValue2 (params, 'reduceOnly', 'reduce_only', false);
+            const openOrClose = (reduceOnly) ? 'close' : 'open';
+            const offset = this.safeString (params, 'offset', openOrClose);
+            request['offset'] = offset;
+            if (reduceOnly) {
+                request['reduce_only'] = 1;
+            }
+            request['lever_rate'] = leverRate;
+            request['order_price_type'] = type;
         }
-        request['order_price_type'] = type;
+        params = this.omit (params, [ 'reduceOnly', 'stopPrice', 'stopLossPrice', 'takeProfitPrice', 'triggerType', 'leverRate' ]);
         const broker = this.safeValue (this.options, 'broker', {});
         const brokerId = this.safeString (broker, 'id');
         request['channel_code'] = brokerId;
-        const clientOrderId = this.safeString2 (params, 'client_order_id', 'clientOrderId');
-        if (clientOrderId !== undefined) {
-            request['client_order_id'] = clientOrderId;
-            params = this.omit (params, [ 'client_order_id', 'clientOrderId' ]);
-        }
-        let method = undefined;
+        let response = undefined;
         if (market['linear']) {
             let marginMode = undefined;
             [ marginMode, params ] = this.handleMarginModeAndParams ('createOrder', params);
             marginMode = (marginMode === undefined) ? 'cross' : marginMode;
             if (marginMode === 'isolated') {
-                method = 'contractPrivatePostLinearSwapApiV1SwapOrder';
+                if (isStop) {
+                    response = await this.contractPrivatePostLinearSwapApiV1SwapTriggerOrder (this.extend (request, params));
+                } else if (isStopLossTriggerOrder || isTakeProfitTriggerOrder) {
```

Comment:
@Dan-krm when I try to place a `takeProfit` order I always get this error, although there is enough margin, do you know why?
```
 p huobi createOrder "LTC/USDT:USDT" limit buy 1 30 '{"takeProfitPrice":31}'          
Python v3.10.9
CCXT v3.0.73
huobi.createOrder(LTC/USDT:USDT,limit,buy,1,30,{'takeProfitPrice': 31})
Traceback (most recent call last):
  File "/Users/cjg/Git/ccxt10/ccxt/examples/py/cli.py", line 232, in <module>
    asyncio.run(main())
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/Users/cjg/Git/ccxt10/ccxt/examples/py/cli.py", line 211, in main
    result = await method(*args)
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/async_support/huobi.py", line 3925, in create_order
    return await getattr(self, method)(symbol, type, side, amount, price, query)
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/async_support/huobi.py", line 4096, in create_contract_order
    response = await self.contractPrivatePostLinearSwapApiV1SwapCrossTpslOrder(self.extend(request, params))
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/async_support/base/exchange.py", line 1613, in request
    return await self.fetch2(path, api, method, params, headers, body, config, context)
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/async_support/base/exchange.py", line 1610, in fetch2
    return await self.fetch(request['url'], request['method'], request['headers'], request['body'])
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/async_support/base/exchange.py", line 180, in fetch
    self.handle_errors(http_status_code, http_status_text, url, method, headers, http_response, json_response, request_headers, request_body)
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/async_support/huobi.py", line 5405, in handle_errors
    self.throw_exactly_matched_exception(self.exceptions['exact'], code, feedback)
  File "/Users/cjg/Git/ccxt10/ccxt/python/ccxt/async_support/base/exchange.py", line 1919, in throw_exactly_matched_exception
    raise exact[string](message)
ccxt.base.errors.InsufficientFunds: huobi {"status":"error","err_code":1048,"err_msg":"Insufficient close amount available.","ts":1682517174654}
```

Reply:
You could be getting this error because Huobi uses a traditional take-profit and stop-loss approach so you need to have an open position to use the take-profit and stop-loss. It doesn't exactly match our current schema because it requires you to make the tp/sl orders separate from the position opening order and you need an open position to make the tp/sl orders

Reply:
@Dan-krm I think I had one opened 🤔

Reply:
@Dan-krm can you place tk/sl orders on your end? if so can you post the logs please?

Reply:
@carlosmiei 
It looks like the issue might be that you need to put sell as the type

```
createOrder BTC/USDT:USDT limit sell 1 30000 '{"takeProfitPrice":29900}'

huobi.createOrder (BTC/USDT:USDT, limit, sell, 1, 30000, [object Object])
fetch Request:
 huobi POST https://api.hbdm.com/linear-swap-api/v1/swap_cross_tpsl_order?AccessKeyId=abc-123&SignatureMethod=HmacSHA256&SignatureVersion=2&Timestamp=2023-04-28T05%3A04%3A42&Signature=abc-123
RequestHeaders:
 {
  'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.75 Safari/537.36',
  'Content-Type': 'application/json'
}
RequestBody:
 {"contract_code":"BTC-USDT","volume":"1","direction":"sell","tp_order_price_type":"limit","tp_trigger_price":"29900","tp_order_price":"30000","channel_code":"abc"}

handleRestResponse:
 huobi POST https://api.hbdm.com/linear-swap-api/v1/swap_cross_tpsl_order?AccessKeyId=abc-123&SignatureMethod=HmacSHA256&SignatureVersion=2&Timestamp=2023-04-28T05%3A04%3A42&Signature=abc-123 200 OK
ResponseHeaders:
 {
  'Cache-Control': 'no-cache, no-store',
  'Cf-Cache-Status': 'DYNAMIC',
  'Cf-Ray': '7becc41c1c1ce27b-ORD',
  Connection: 'keep-alive',
  'Content-Encoding': 'gzip',
  'Content-Type': 'text/plain;charset=UTF-8',
  Date: 'Fri, 28 Apr 2023 05:04:43 GMT',
  Expires: 'Thu, 01 Jan 1970 00:00:00 GMT',
  Pragma: 'no-cache',
  'Ratelimit-Interval': '1000',
  'Ratelimit-Limit': '5',
  'Ratelimit-Remaining': '4',
  'Ratelimit-Reset': '1682658284013',
  Reqid: '90f7abac217c3db922d31317b6c748ae',
  'Request-Id': '022afde4007dc0593ac4baf28f8dde7d',
  Server: 'cloudflare',
  'Transfer-Encoding': 'chunked',
  'X-Request-Id': '90f7abac217c3db922d31317b6c748ae'
}
ResponseBody:
 {"status":"ok","data":{"tp_order":{"order_id":1101494204040163328,"order_id_str":"1101494204040163328"},"sl_order":null},"ts":1682658283024}

2023-04-28T05:04:42.849Z iteration 0 passed in 419 ms

{
  info: {
    tp_order: {
      order_id: '1101494204040163328',
      order_id_str: '1101494204040163328'
    },
    sl_order: null
  },
  id: undefined,
  clientOrderId: undefined,
  timestamp: undefined,
  datetime: undefined,
  lastTradeTimestamp: undefined,
  symbol: 'BTC/USDT:USDT',
  type: undefined,
  timeInForce: undefined,
  postOnly: undefined,
  side: undefined,
  price: undefined,
  stopPrice: undefined,
  triggerPrice: undefined,
  average: undefined,
  cost: undefined,
  amount: undefined,
  filled: undefined,
  remaining: undefined,
  status: undefined,
  fee: undefined,
  trades: [],
  fees: [],
  reduceOnly: undefined
}
```

Reply:
@Dan-krm you're right for clarifying it!